File: src/BlazorStrap/util/BlazorStrapInterop.cs
using System.Threading.Tasks;
using Microsoft.AspNetCore.Components;
using Microsoft.JSInterop;

namespace BlazorStrap.util
{
    public class BlazorStrapInterop
    {
        private IJSRuntime JSRuntime { get; set; }

        public BlazorStrapInterop(IJSRuntime jsRuntime)
        {
            JSRuntime = jsRuntime;
        }

        public Task<bool> ChangeBody(string classname)
            => JSRuntime.InvokeAsync<bool>("blazorStrap.changeBody", classname);

        public Task<bool> Log(string message)
            => JSRuntime.InvokeAsync<bool>("blazorStrap.log", message);

        public Task<bool> Popper(string target, string popper, ElementRef arrow, string placement)
            => JSRuntime.InvokeAsync<bool>("blazorStrap.popper", target, popper, arrow, placement);

        public Task<bool> Tooltip(string target, ElementRef tooltip, ElementRef arrow, string placement)
            => JSRuntime.InvokeAsync<bool>("blazorStrap.tooltip", target, tooltip, arrow, placement);

        public Task FocusElement(ElementRef el) 
            => JSRuntime.InvokeAsync<object>("blazorStrap.focusElement", el);
    }
}

File: src/BlazorStrap/content/blazorStrapInterop.js
Blazor.registerFunction('BlazorStrap.BlazorStrapInterop.Log', function (message) {
    console.log('message: ', message);
    return true;
});

Blazor.registerFunction('BlazorStrap.BlazorStrapInterop.ChangeBody', function (classname) {
    document.body.className = classname;
    return true;
});

Blazor.registerFunction('BlazorStrap.BlazorStrapInterop.Popper', function (target, popperId, arrow, placement) {
    var reference = document.getElementById(target);
    var popper = document.getElementById(popperId);
    showPopper(reference, popper, arrow, placement);
    return true;
});

Blazor.registerFunction('BlazorStrap.BlazorStrapInterop.Tooltip', function (target, tooltip, arrow, placement) {
    var reference = document.getElementById(target);
    reference.addEventListener('mouseover', function () {
        tooltip.className = "tooltip fade show bs-popover-" + placement;
        showPopper(reference, tooltip, arrow, placement);
    });
    reference.addEventListener('mouseout', function () {
        tooltip.className = "tooltip hide";
    });
    return true;
});

function showPopper(reference, popper, arrow, placement) {
    var thePopper = new Popper(reference, popper,
        {
            placement,
            modifiers: {
                offset: {
                    offset: 0
                },
                flip: {
                    behavior: 'flip'
                },
                arrow: {
                    element: arrow,
                    enabled: true
                },
                preventOverflow: {
                    boundary: 'scrollParent'
                }

            }
        }
    );
}
